### PR TITLE
[CodeGen] change prototype of regalloc filter function

### PR DESCRIPTION
**Pull request by cdevadas**
[CodeGen] Change prototype of regalloc filter function

Change the prototype of the filter function so that we can
filter not just by RegClass. We need to implement more
complicated filter based upon some other info associated
with each register.

Patch provided by: Gang Chen (gangc@amd.com)

### CODE REVIEW

**Comment by cdevadas** [comment]
The use case for the altered filter function is in the follow-up patch https://github.com/llvm/llvm-project/pull/93526.

**Comment by cdevadas** [comment]
The filtering is now more on the Registers itself than the regclass; hence, the following changes make more sense. If everyone thinks it is a good idea, I can change it.
RegClassFilterFunc -> RegAllocFilterFunc
ShouldAllocateClass -> ShouldAllocateRegister

File: llvm/include/llvm/CodeGen/RegAllocCommon.h [hunk 1]
```diff
@@ -10,22 +10,25 @@
 #define LLVM_CODEGEN_REGALLOCCOMMON_H
 
 #include <functional>
+#include <llvm/CodeGen/Register.h>
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/RegAllocCommon.h:13]: Should use `""` instead of `<>` for project local includes

**Review by arsenm — APPROVED** [review]
LGTM, but wait for the other patches to submit first

**Comment by cdevadas** [comment]
Fixed the header inclusion and renamed the filter function name.

**Review by cmc-rep — APPROVED** [review]

**Pull request merged**